Category custom fields marked as 'subform' act as if they are not

nikosdion opened the pull request:
Pull Request for Issue #35543 .

### Summary of Changes

Category custom fields marked as “Use Only In Subform” were loaded as regular fields in all categories. This is a two–part bug (see Technical Notes at the bottom).

As a result I made the following changes:
* I changed `\Joomla\Component\Fields\Administrator\Helper\FieldsHelper::getFields` to only load non–subform–only fields by default with a new option to return all fields, including the subform–only, and made the necessary changes to SubfieldsField and PlgFieldsSubform.
* I changed PlgSystemFields::onContentPrepareForm to load the missing form data when you are saving a com_content category.

### Testing Instructions

* In the backend of your site go to Content, Fields.
* From the dropdown select “Category” and click on New.
* Create a new field with the following parameters:
   - Title: Test
   - Label: Test
   - Required: Yes. **THIS IS IMPORTANT**
   - Only Use In Subform: Yes. **THIS IS IMPORTANT**
   - Filter: Safe HTML. **THIS IS IMPORTANT**
* Click on Save & Close
* Go to Content, Categories
* Edit _any category_
* Click on Save

### Actual result BEFORE applying this Pull Request

You get an error that the Test field is required.

### Expected result AFTER applying this Pull Request

No error; the category is saved correctly.

### Documentation Changes Required

None. I mean, it would be nice if someone documented the “Only Use In Subform” field as now there's no tooltip and no documentation. What does this do? Nobody knows. Even looking at the code it was not obvious, presumably because it did nothing at all to begin with...

### Backwards compatibility notes

The `\Joomla\Component\Fields\Administrator\Helper\FieldsHelper::getFields` method has changed so that, by default, it does not return ALL fields but only those not marked as “Only Use In Subform”. Is this a b/c break? I would say no, because the intent of this option was to make these fields unavailable anywhere EXCEPT the subform custom field. Fixing a bug so that something works as it should, per the meager code documentation, shouldn't be considered a b/c break.

If it is to be considered a b/c break then the “Only Use In Subform“ feature should be removed from 4.0 as it cannot be fixed otherwise (well, just part II of the fix will kinda–sorta fix it but it's an ancillary artefact at best). Then again, removing a feature — even one that never worked — would be a b/c break.

So I am going to ping @wilsonge and @zero-24 and let them sort out the decision–making process of what constitutes a backward compatibility break in this context.

### Technical notes

This is a two part issue.

#### Part I. “Only Use In Subform” doesn't do anything

The new “Only Use In Subform” option in a Custom Field's definition not doing what it's supposed to do.

Supposedly, custom fields with this option enabled would be excluded from everything _except_ defining or rendering a subform custom field.

In reality, this option does nothing useful. It displays a badge in the fields list. Oh, yeah, it also makes the Custom Fields where this option is enabled have a category list of `[-1]`. That's the beginning, middle and end of it. Totally useless.

I changed `\Joomla\Component\Fields\Administrator\Helper\FieldsHelper::getFields` to only load non–subform–only fields by default. If you want to get all fields, including subform–only, you need to pass `true` to a new fifth parameter. Note that the ONLY use case for retrieving all fields, including subform–only, is the SubfieldsField and the subform field itself. Third–party software SHOULD NOT receive those fields outside of the context of a subform custom field, therefore this is not a b/c break.

#### Part II. `PlgSystemFields::onContentPrepareForm` doesn't actually work for com_categories (when saving)

You will see that the system plugin **ALLEGEDLY** has special handling for editing categories in com_categories. Allegedly, because it works on the edit page but NOT when saving! When saving all fields from all categories are loaded since the plugin runs before the form data is loaded from the request, meaning there is no category filter present.

You see, `$data` is an empty array when you the `onContentPrepareForm` event is fired _when saving the form_. I see that  the choice to not populate the form data when loading the form in `\Joomla\CMS\MVC\Controller\FormController::save` dates back to 2010 when Andrew made a commit about fixing some unspecified issue saving plugins.

The problem is that for com_categories _and only this component_ we actually need to load the form data from the request when saving data. Otherwise PlgSystemFields:: onContentPrepareForm sees that $data is empty and doesn't set the category ID. As a result `FieldsHelper::prepareForm` populates `$assignedCatids` to null, therefore it loads all fields from all categories. If any of these fields is required AND has a validator which doesn't accept NULL values — like the ones @crystalenka mentioned — you get a save error.

There are a few options to fix it. One is to override CategoryController::save. However, this results in severe code duplication which will be a pain in the posterior to manage whenever the base FormController::save method is updated.

The other option is to force CategoryModel::getForm to always load the data. However, this may break b/c; I'm especially concerned about third party plugins which may use the empty $data as a signal when figuring out which fields to add to the category edit view.

The final option is to change PlgSystemFields::onContentPrepareForm itself. Since we're doing special handling for com_categories we might just as well populate `data` from the request if we are saving the form. This is what I plan on doing. To keep things b/c this nasty workaround will ONLY apply to com_content categories, not third party extensions' categories. In fact, custom fields SHOULD only apply to com_content categories — which is not the case now — but that's a different bug for a different day, _n'est ce pas_?

Commit:
Subform–only fields cause saving a category to fail.

Part I: “Only Use In Subform” doesn't do anything

Commit:
Subform–only fields cause saving a category to fail.

Part II. PlgSystemFields::onContentPrepareForm doesn't actually work for com_categories

Close gh-35543

Commit:
Subform–only fields cause saving a category to fail.

Part II. PlgSystemFields::onContentPrepareForm doesn't actually work for com_categories

This should only apply on com_content categories, not
third party categories.

richard67 commented on administrator/components/com_fields/src/Helper/FieldsHelper.php:
```diff
@@ -108,6 +108,15 @@ public static function getFields($context, $item = null, $prepareValue = false,
 			self::$fieldsCache->setState('list.limit', 0);
 		}
 
+		if (!$includeSubformFields)
+		{
+			self::$fieldsCache->setState('filter.only_use_in_subform', 0);
+		}
+		else
+		{
+			self::$fieldsCache->setState('filter.only_use_in_subform', '');
+		}
```
```suggestion
		if ($includeSubformFields)
		{
			self::$fieldsCache->setState('filter.only_use_in_subform', '');
		}
		else
		{
			self::$fieldsCache->setState('filter.only_use_in_subform', 0);
		}
```
Couldn't we use positive logic here?

nikosdion replied:
Either way works for me 🤷🏽‍♂️

richard67 replied:
It's nicer with positive logic according to school books ;-)

nikosdion replied:
I prefer to have the code that applies the non–default filter first. This makes it far less likely that someone will see it and go “oh, this parameter and if–block are redundant, we are just setting the filter to its default value”, remove it and break stuff because they didn't actually read the code or run a git blame to see WTF this code is there. It may or may not have happened before...

nikosdion replied:
BTW I did merge your suggestion or at least I told GitHub to do so.

richard67 replied:
Yes, you merged it. But if you think it was wrong and your original code was better to understand or to read, then change it back. It was just a suggestion, and as a human being I might be wrong with a suggestion sometimes. It's just school books say positive logic shall be preferred because normally that's better readable.

nikosdion replied:
Nah, it's all good. Let's keep it the way you proposed. It's probably more readable for people who are not like me.

Being neurodiverse I read code differently than most people. I don't care about inverted or negative logic. Same for double negatives. They don't trip me over. What trips me over to some extent is if the “interesting“ branch of an if–block is not at the top. It slows me down reading that code. I also recognise that this is NOT how most people read code — I taught other people to code about a decade ago, a very informing and humbling experience for me. So, when it's not code that I am most likely to mostly read and work on myself I am happy to accommodate the majority of people who will be reading and working on it. In the end of the day FOSS is about helping _other people_ with our code, right? 😄

Commit:
Invert the if–block order

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Commit:
Subform–only fields cause saving a category to fail.

Part II. PlgSystemFields::onContentPrepareForm doesn't actually work for com_categories

Okay, it could actually apply for any category,
not just com_content.

crystalenka commented:
I have tested this item :white_check_mark: successfully on 5879d5690ec4e91665e75c80e5e195a85c434e51<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35547">issues.joomla.org/tracker/joomla-cms/35547</a>.</sub>

Fedik commented on administrator/components/com_fields/src/Helper/FieldsHelper.php:
```diff
@@ -108,6 +108,15 @@ public static function getFields($context, $item = null, $prepareValue = false,
 
+		if ($includeSubformFields)
+		{
+			self::$fieldsCache->setState('filter.only_use_in_subform', '');
```
Should not this be `1` instead of empty `''`? 
https://github.com/joomla/joomla-cms/blob/d151fd31af85b93fc84378c1bd1643d60bd03b47/administrator/components/com_fields/src/Model/FieldsModel.php#L352-L359

The field `only_use_in_subform` itself is a radio button with `0` and `1`

nikosdion replied:
No. If you use 1 it will return **only** the fields set up as only use in subform. Subforms must have access to all fields, no matter if they were marked as only in subform or not. That's why I use an empty string.

Everywhere else we must ONLY return the fields which are NOT subform only. This is why in the other branch of the if-block I set the filter to 0.

Remember that all publish-style fields which accept an integer will filter records where the respective field is this integer. However, they also have an empty state. When the field is an empty string no filtering will be applied.

Think about a published filter field. If it's an empty string you show all records, published or unpublished. If it's 0 you show unpublished records. If it's 1 you show published records. The same thing applies here.

Are we seriously discussing how Joomla has been working since it was called Mambo...?

Fedik commented:
Looks good to me.

The `filter.only_use_in_subform` was add in #33096 , 
@joomdonation can you please have a look, I think  you forgot to add the `set('filter.only_use_in_subform')` for it or?
The Article does not have this bug because the fields always filtered by `filter.assigned_cat_ids`, however for category this filter always empty.

Fedik commented:
I have tested this item :white_check_mark: successfully on 5879d5690ec4e91665e75c80e5e195a85c434e51<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35547">issues.joomla.org/tracker/joomla-cms/35547</a>.</sub>

Fedik commented on administrator/components/com_fields/src/Helper/FieldsHelper.php:
```diff
@@ -96,7 +96,7 @@ public static function extract($contextString, $item = null)
 	 * @since   3.7.0
 	 */
-	public static function getFields($context, $item = null, $prepareValue = false, array $valuesToOverride = null)
+	public static function getFields($context, $item = null, $prepareValue = false, array $valuesToOverride = null, bool $includeSubformFields = false)
```
Please update doc block. Add description  for a new parameter

nikosdion replied:
Done!

Commit:
Docblock

Fedik commented:
RTC
It had 2 tests, and last change was a comment correction<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35547">issues.joomla.org/tracker/joomla-cms/35547</a>.</sub>

Fedik commented:
@wilsonge still have to decide when where and whether and  merge it ;)

wilsonge commented:
Joyfully since PHP 8 this counts as a b/c break because method signatures aren't allowed to change if people are subclassing example: https://3v4l.org/iedIW . This is also causing issues with other PRs such as https://github.com/joomla/joomla-cms/pull/35476 (where I stole the example link from).

I think it's fairly unlikely that people are subclassing fieldshelper to be honest so my gut is to merge this. However I'm going to leave this here a few days to see if anyone else has any strong objections or feelings on the matter.

nikosdion commented:
I'd be surprised if anyone subclasses that. The way it's constructed it's just a magic static helper you can't really modify without hacking core — or making a PR :trollface:

wilsonge commented:
No complaints imminent. Merged. Thanks!

nikosdion commented:
Thank you!

janschoenherr commented:
Hi, can someone please review this PR: https://github.com/joomla/joomla-cms/pull/35924

AndySDH commented:
Thank you @nikosdion for making this, a great fix! This also fixed the same issue for the Users component! (as they have no category either, so they ran into the same problem)
